### PR TITLE
Path creater

### PR DESCRIPTION
**Pull request by elproblema**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.19)
+project(disj_paths_bd_tw)
+
+set(CMAKE_CXX_STANDARD 14)
+
+include_directories("/home/tunyash/boost_1_76_0/")
```
Comment (tunyash) [inline CMakeLists.txt]: Change to include_directories(~/boost_1_76_0/)

File: src/path-decomposition.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        _error_type(INCONTINUITY), _bag_i(i), _bag_j(j), _bag_k(k), _violating_node(u) {}
+
+const char *PathDecomposition::CorectnessException::what() const throw() {
+    switch (_error_type) {
+        case EDGES:
+            return "Not all edges are in path-width decomposition";
+        case NEGATIVE_VERTICES:
+            return "Vertices in bags are not in [0; |V(_g)|)";
+        case INCONTINUITY:
+            return ("_bags["
```
Comment (tunyash) [inline src/path-decomposition.cpp]: It doesn't seem to be safe to use c_str() like that. You create a std::string which gets deleted after you exit the method and you return a pointer to the start of this string which may no longer be valid. Store the std::string as a field in the class instead.

File: src/path-decomposition.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    std::vector<std::vector<int>> size(n, std::vector<int>(log_n + 2));
+    // size[v][d] = size of subtree of vertex |v| on depth |d|
+    std::vector<int> used(n);
+    // used[v] = 1 if vertex |v| is already in centroid tree
+    // used[v] = 0 else
+    std::vector<int> parent(n, -1);
+    // parent[v] = parent of vertex |v| in centroid tree
+
+    std::function<void(int, int, int)> size_calculation;
+    size_calculation = [&](tree_vertex v, int depth, tree_vertex p) {
```
Comment (tunyash) [inline src/path-decomposition.cpp]: I think this is too complicated to be inside a method. This makes it untestable separately which is not ideal given how error-prone tdlib is. Make a separate function that accepts a tree decomposition and returns a path decomposition.

File: src/path-decomposition.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+}
+
+PathDecomposition::PathDecomposition(std::vector<std::vector<vertex_t>> bags, Graph g): _bags(bags), _g(g) {
+    Check();
+}
+
+PathDecomposition::PathDecomposition(Graph g): _g(g) {
+    transform();
+    Check();
```
Comment (tunyash) [inline src/path-decomposition.cpp]: decapitalize, the naming convention for methods is lowercase words separated by subscripts.

File: src/test.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    } catch(...) {
+        pass = false;
+    }
+    CHECK(pass);
+}
+
+TEST_CASE("grid complete graph") {
+    srand(2281337);
+
+    int n = 30;
```
Comment (tunyash) [inline src/test.cpp]: Make it small, so unit tests can pass quickly.

File: src/test.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    bool pass = true;
+    try {
+        PathDecomposition p(g);
+    } catch(...) {
+        pass = false;
+    }
+    CHECK(pass);
+}
+
+TEST_CASE("grid complete graph") {
```
Comment (tunyash) [inline src/test.cpp]: "complete grid"

File: src/test.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    bool pass = true;
+    try {
+        PathDecomposition p(g);
+    } catch(...) {
+        pass = false;
+    }
+    CHECK(pass);
+}
+
+TEST_CASE("grid random graph") {
```
Comment (tunyash) [inline src/test.cpp]: "sparse grid"

File: src/path-decomposition.h [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        /*
+        This is used only for error_type == 2
+        @error_type == 2 used when _bags[i] and _bags[k] contains vertex |v| but _bags[j] does not.
+        i < j < k
+        */
+        ~CorectnessException() throw() {}
+        const char* what() const throw();
+    };
+
+    PathDecomposition(std::vector<std::vector<vertex_t>> bags, Graph g);
```
Comment (tunyash) [inline src/path-decomposition.h]: Pass as a link or use std::move

File: src/path-decomposition.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        int _error_type;
+        int _bag_i, _bag_j, _bag_k, _violating_node;
+
+        CorectnessException(int error_type);
+        /*
+        @error_type == 0 used when in graph |_g| exists edge (v, u) but in |_bags| there is no bag that contains {v, u}
+        @error_type == 1 used when in |_bags| exists vertex |v| that does not belong to [0; |V(_g)|)
+        @error_type == 3 used when exists vertex |v| that does not belong to |_bags|
+        */
+        CorectnessException(int i, int j, int k, int u);
```
Comment (tunyash) [inline src/path-decomposition.h]: what is u? i -> bag_i (for j and k as well)

File: src/path-decomposition.h [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    const Graph& get_graph() const {
+        return _g;
+    }
+
+
+private:
+
+    void Check();
```
Comment (tunyash) [inline src/path-decomposition.h]: decapitalize

**Review by tunyash — CHANGES_REQUESTED** [review]
Rollback tdlib changes as well.

File: src/path-decomposition.h (hunk 9 above)
Comment (elproblema) [reply src/path-decomposition.h]: There is a comment below

File: src/path-decomposition.cpp (hunk 3 above)
Comment (elproblema) [reply src/path-decomposition.cpp]: I think that never-used function will clutter up path-decomposition.h interface

File: src/path-decomposition.h (hunk 9 above)
Comment (tunyash) [reply src/path-decomposition.h]: There is no u in the comment. Please give the arguments more meaningful names.

File: src/path-decomposition.cpp (hunk 3 above)
Comment (tunyash) [reply src/path-decomposition.cpp]: given the issues of tdlib, the testability of this algo separately is more important.

File: src/path-decomposition.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        std::sort(bag.begin(), bag.end());
+        bag.resize(std::unique(bag.begin(), bag.end()) - bag.begin());
+        //Erasing non-unique vertices from bags
+        for (int v : bag) {
+            if (v < 0 || v >= n)
+                throw PathDecomposition::CorectnessException(NEGATIVE_VERTICES);
+        }
+    }
+
+    std::vector<int> sum(n);
```
Comment (tunyash) [inline src/path-decomposition.cpp]: sum, left, and right are very generic and non-descriptive names, I'd prefer something like containing_bags_count, left_bag, right_bag instead.

File: src/centroid-tree.h [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+        TreeException(Tree &t);
+        /*
+         * @param |t| is graph where a loop was found
+         */
+
+        const char * what() const throw();
+    };
+
+    CentroidTree(Tree &t);
```
Comment (tunyash) [inline src/centroid-tree.h]: explicit

File: src/centroid-tree.h [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    /*
+     * @param |t| is tree-width decomposition of some graph |g|
+     * P.S.: same graph |g| is used in the get_path_decomposition() function
+     */
+
+    CentroidTree(Graph &g);
```
Comment (tunyash) [inline src/centroid-tree.h]: explicit

File: src/centroid-tree.h [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    class TreeException : public std::exception {
+    /*
+     * Thrown when |_tree| is not a tree
+     */
+
+    public:
+
+        Tree &_t;
+
+        TreeException(Tree &t);
```
Comment (tunyash) [inline src/centroid-tree.h]: explicit

File: src/test.cpp [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    } catch(...) {
+        pass = false;
+    }
+    CHECK(pass);
+}
+
+TEST_CASE("sparse grid") {
+    srand(2281337);
+
+    int n = 20;
```
Comment (tunyash) [inline src/test.cpp]: let's decrease this one

File: src/centroid-tree.cpp [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    int log_n = 0, n = boost::num_vertices(_tree);
+    while ((1 << log_n) < n) ++log_n;
+    std::vector<std::vector<int>> size(n, std::vector<int>(log_n + 2));
+    // size[v][d] = size of subtree of vertex |v| on depth |d|
+    std::vector<int> used(n);
+    // used[v] = 1 if vertex |v| is already in centroid tree
+    // used[v] = 0 else
+
+    std::function<void(int, int, int)> size_calculation;
+    size_calculation = [&](tree_vertex_t v, int depth, tree_vertex_t p) {
```
Comment (tunyash) [inline src/centroid-tree.cpp:49]: Please make these std::functions private funcitons of the class

File: src/centroid-tree.cpp [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    return PathDecomposition(bags, g);
+}
+
+CentroidTree::CentroidTree(Graph &g) {
+    Tree t;
+    treedec::exact_decomposition_cutset(g, t);
+    *this = CentroidTree(t);
+}
+
+CentroidTree::CentroidTree(Tree &t): _tree(t) {
```
Comment (tunyash) [inline src/centroid-tree.cpp:120]: Please add a unit test independent of tdlib (e.g. generate a tree-decomposition of a grid and pass it to this constructor).

**Review by tunyash — APPROVED** [review]
LGTM, given my comments are addressed.

**Comment by elproblema** [comment]
Make unit-tests running without errors

**Pull request merged**